pkg/yang/number_test.go

```diff
@@ -15,322 +15,338 @@
+		n2 = maxNumber
+	}
+	return YRange{n1, n2}
+}
+
+func TestRangeEqual(t *testing.T) {
+	for x, tt := range []struct {
+		r1, r2 YangRange
+		ok     bool
+	}{
+		{ok: true},                          // empty range contained in empty range
+		{r1: YangRange{R(1, 2)}, ok: false}, // empty range contained in range
+		{r2: YangRange{R(1, 2)}, ok: false}, // range contained in empty range
+		{
+			YangRange{R(1, 2)},
+			YangRange{R(1, 2)},
+			true,
+		},
+		{
+			YangRange{R(1, 3)},
+			YangRange{R(1, 2)},
+			false,
+		},
+		{
+			YangRange{R(1, 2), R(4, 5)},
+			YangRange{R(1, 2), R(4, 5)},
+			true,
+		},
+		{
+			YangRange{R(1, 2), R(4, 6)},
+			YangRange{R(1, 2), R(4, 5)},
+			false,
+		},
+		{
+			YangRange{R(1, 2)},
+			YangRange{R(1, 2), R(4, 5)},
+			false,
+		},
+		{
+			YangRange{R(1, 2), R(4, 5)},
+			YangRange{R(1, 2)},
+			false,
+		},
+	} {
+		if ok := tt.r1.Equal(tt.r2); ok != tt.ok {
+			t.Errorf("#%d: got %v, want %v", x, ok, tt.ok)
+		}
+	}
+}
+
+func TestRangeContains(t *testing.T) {
+	for x, tt := range []struct {
+		r1, r2 YangRange
+		ok     bool
+	}{
+		{ok: true},
+		{r1: YangRange{R(1, 2)}, ok: true},
+		{r2: YangRange{R(1, 2)}, ok: true},
+		{
+			r1: YangRange{R(1, 2)},
+			r2: YangRange{R(1, 2)},
+			ok: true,
+		},
+		{
+			r1: YangRange{R(1, 5)},
+			r2: YangRange{R(2, 3)},
+			ok: true,
+		},
+		{
+			r1: YangRange{R(2, 3)},
+			r2: YangRange{R(1, 5)},
+			ok: false,
+		},
+		{
+			r1: YangRange{R(1, 10)},
+			r2: YangRange{R(1, 2), R(4, 5), R(7, 10)},
+			ok: true,
+		},
+		{
+			r1: YangRange{R(1, 10)},
+			r2: YangRange{R(1, 2), R(7, 11)},
+			ok: false,
+		},
+		{
+			r1: YangRange{R(1, 9), R(11, 19), R(21, 29)},
+			r2: YangRange{R(23, 25)},
+			ok: true,
+		},
+		{
+			r1: YangRange{R(1, 9), R(11, 19), R(21, 29)},
+			r2: YangRange{R(23, 23)},
+			ok: true,
+		},
+		{
+			r1: YangRange{R(1, 9), R(11, 19), R(21, 29)},
+			r2: YangRange{R(20, 20)},
+			ok: false,
+		},
+		{
+			r1: YangRange{R(1, 10)},
+			r2: YangRange{R(useMin, useMax)},
+			ok: true,
+		},
+		{
+			r1: YangRange{R(useMin, useMax)},
+			r2: YangRange{R(1, 10)},
+			ok: true,
+		},
+		{
+			r1: YangRange{R(1024, 65535)},
+			r2: YangRange{R(useMin, 4096), R(5120, useMax)},
+			ok: true,
+		},
+		{
+			r1: YangRange{R(1024, 65535)},
+			r2: YangRange{R(-999999, 4096), R(5120, useMax)},
+			ok: false,
+		},
+		{
+			r1: YangRange{R(1024, 65535)},
+			r2: YangRange{R(useMin, 4096), R(5120, 999999)},
+			ok: false,
+		},
+	} {
+		if ok := tt.r1.Contains(tt.r2); ok != tt.ok {
+			t.Errorf("#%d: got %v, want %v", x, ok, tt.ok)
+		}
+	}
 }
 
-// errToStr outputs e's error string if it is not-nil, or an empty string
-// otherwise.
-func errToStr(e error) string {
-	if e == nil {
-		return ""
+func TestCoalesce(t *testing.T) {
+	for x, tt := range []struct {
+		in, out YangRange
+	}{
+		{},
+		{YangRange{R(1, 4)}, YangRange{R(1, 4)}},
+		{YangRange{R(1, 2), R(3, 4)}, YangRange{R(1, 4)}},
+		{YangRange{R(1, 2), R(2, 4)}, YangRange{R(1, 4)}},
```

Comment:
coalesce is still treating the range as an integer range when coalescing, so it's erroneously coalescing decimal ranges:

TestCoalesce:
```
                {
                        YangRange{YRange{Min: FromFloat(1), Max: FromFloat(2.5)}, YRange{Min: FromFloat(3), Max: FromFloat(4)}},
                        YangRange{YRange{Min: FromFloat(1), Max: FromFloat(2.5)}, YRange{Min: FromFloat(3), Max: FromFloat(4)}},
                },
```
number_test.go:187: #3: got 1.0..4.0, want 1.0..2.5|3.0..4.0

We can fix by either
1. support decimal range coalescing by adding the fractional digit instead of add(1).
2. not coalesce for decimal ranges.

We can do either by adding to `ParseRanges` a YANG typename parameter or have two `ParseRanges` for ints and `decimal64`. If the latter, then we can optionally change the current `TestCoalesce` to `TestParseRanges` as maybe that's the better test to have (or I'd be happy to do it in a follow-up PR).

Reply:
decided to support decimal range coalescing by splitting ParseRanges into a variation for int and decimal.